### PR TITLE
feat: add group of events feature

### PR DESCRIPTION
**Pull request by Tlazypanda**
Fixes #

#### Short description of what this resolves:
Add minimal group of events feature. 

#### Checklist

- [ ] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [ ] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by Tlazypanda** [comment]
@iamareebjamal can you please review? :sweat_smile: I wasn't sure what permissions to set for the api so have just added minimal ones for now please let me know about this

File: app/models/group.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from app.models import db
+from app.models.base import SoftDeletionModel
+
+
+class Group(SoftDeletionModel):
+    __tablename__ = 'groups'
+
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String, nullable=False)
+    user_id = db.Column(db.Integer, db.ForeignKey('users.id', ondelete='CASCADE'))
```
Comment (iamareebjamal) [inline app/models/group.py:12]: This should be nullable False, we always need a user for a group

File: migrations/versions/rev-2021-01-13-12:50:33-8277175ec726_.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('name', sa.String(), nullable=False),
+    sa.Column('user_id', sa.Integer(), nullable=True),
+    sa.Column('created_at', sa.DateTime(timezone=True), nullable=True),
+    sa.Column('modified_at', sa.DateTime(timezone=True), nullable=True),
+    sa.ForeignKeyConstraint(['user_id'], ['users.id'], ondelete='CASCADE'),
+    sa.PrimaryKeyConstraint('id')
+    )
+    op.add_column('events', sa.Column('group_id', sa.Integer(), nullable=True))
+    op.create_foreign_key(None, 'events', 'groups', ['group_id'], ['id'], ondelete='SET NULL')
```
Comment (iamareebjamal) [inline migrations/versions/rev-2021-01-13-12:50:33-8277175ec726_.py:32]: Add name to the foreign key here, so it becomes possible to drop it below

See other migrations for reference

File: migrations/versions/rev-2021-01-13-12:50:33-8277175ec726_.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    op.add_column('events_version', sa.Column('group_id', sa.Integer(), autoincrement=False, nullable=True))
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_column('events_version', 'group_id')
+    op.drop_constraint(None, 'events', type_='foreignkey')
```
Comment (iamareebjamal) [inline migrations/versions/rev-2021-01-13-12:50:33-8277175ec726_.py:40]: Use the same foreign key name used above here as well

File: app/api/groups.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        """
+        method to check for required relationship with group
+        :param args:
+        :param kwargs:
+        :param data:
+        :return:
+        """
+        data['user'] = current_user.id
+        user = User.query.get(data['user'])
+        if not user.is_verified:
```
Comment (iamareebjamal) [inline app/api/groups.py:32]: You already have current_user, no reason to query user

File: app/api/groups.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        query method for GroupList class
+        :param view_kwargs:
+        :return:
+        """
+        query_ = self.session.query(Group)
+
+        if view_kwargs.get('user_id') and 'GET' in request.method:
+            if not has_access('is_user_itself', user_id=view_kwargs['user_id']):
+                raise ForbiddenError({'source': ''}, 'Access Forbidden')
+            user = safe_query_kwargs(User, view_kwargs, 'user_id')
```
Comment (iamareebjamal) [inline app/api/groups.py:59]: Why not just use view_kwargs['user_id'] in query below instead of querying user

File: app/api/groups.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        :return:
+        """
+        group = safe_query_kwargs(Group, view_kwargs, 'id')
+        if not has_access(
+            'is_user_itself',
+            user_id=group.user_id,
+        ):
+            raise ForbiddenError(
+                {'source': 'User'}, 'You are not authorized to access this.'
+            )
```
Comment (iamareebjamal) [inline app/api/groups.py:92]: Groups are public, they can be queried by anyone

File: app/api/groups.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
```
Comment (iamareebjamal) [inline app/api/groups.py:158]: Groups are public

File: app/api/routes.py [hunk 8]
```diff
@@ -269,6 +270,7 @@
     '/attendees/<int:attendee_id>/user',
     '/feedbacks/<int:feedback_id>/user',
     '/events/<int:event_id>/owner',
+    '/groups/<int:group_id>/owner',
```
Comment (iamareebjamal) [inline app/api/routes.py:273]:
```suggestion
    '/groups/<int:group_id>/user',
```

File: app/api/routes.py [hunk 9]
```diff
@@ -1374,6 +1377,25 @@
...
+)
+api.route(
+    GroupRelationship,
+    'group_events',
+    '/groups/<int:id>/relationships/events',
+)
+api.route(
+    GroupRelationship,
+    'group_owner',
+    '/groups/<int:id>/relationships/owner',
```
Comment (iamareebjamal) [inline app/api/routes.py:1396]:
```suggestion
    '/groups/<int:id>/relationships/user',
```

File: app/api/schema/groups.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        self_view='v1.group_events',
+        self_view_kwargs={'id': '<id>'},
+        related_view='v1.event_list',
+        related_view_kwargs={'group_id': '<id>'},
+        many=True,
+        schema='EventSchemaPublic',
+        type_='event',
+    )
+
+    owner = Relationship(
```
Comment (iamareebjamal) [inline app/api/schema/groups.py:37]:
```suggestion
    user = Relationship(
```

**Comment by iamareebjamal** [comment]
@Tlazypanda Great work!

**Comment by Tlazypanda** [comment]
@iamareebjamal is it okay now? Although the migration number I think has some issue 😅

**Comment by iamareebjamal** [comment]
Yes, rebase the branch with development and update the down revision to correct number

**Comment by Tlazypanda** [comment]
@iamareebjamal  is it okay now? :sweat_smile:

File: app/api/groups.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        },
+    }
+
+
+class GroupRelationship(ResourceRelationship):
+    """
+    Group Relationship
+    """
+
+    decorators = (jwt_required,)
```
Comment (iamareebjamal) [inline app/api/groups.py:126]: Not required to be logged in for getting Group

File: app/api/groups.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        group = safe_query_kwargs(Group, view_kwargs, 'id')
+        if not has_access(
+            'is_user_itself',
+            user_id=group.user_id,
+        ):
+            raise ForbiddenError(
+                {'source': 'User'}, 'You are not authorized to access this.'
+            )
+
+    decorators = (jwt_required,)
```
Comment (iamareebjamal) [inline app/api/groups.py:109]: Not required to be logged in for getting Group

**Comment by iamareebjamal** [comment]
Great! Thank you. These APIs are remaining I believe

/groups/ID/events
/events/ID/group

Also, group needs to be added in event schema

But we can take these in follow up PR if you want?

**Comment by Tlazypanda** [comment]
@iamareebjamal yes would prefer to take this up in a new pr since i have to again set the migrations to test it out :sweat_smile:

**Pull request merged**